cmd/xl-sets.go

```diff
@@ -0,0 +1,495 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/hash"
+	"github.com/minio/minio/pkg/sync/errgroup"
+)
+
+// xlSets implements ObjectLayer combining a static list of erasure coded
+// object layers. NOTE: There is no dynamic scaling allowed or intended in
+// current design.
+type xlSets struct {
+	layers []*xlObjects
+
+	// Pack level listObjects pool management.
+	listPool *treeWalkPool
+}
+
+// Initialize new set of erasure coded sets.
+func newXLSets(layers []*xlObjects) ObjectLayer {
+	return &xlSets{
+		layers:   layers,
+		listPool: newTreeWalkPool(globalLookupTimeout),
+	}
+}
+
+// StorageInfo - combines output of StorageInfo across all erasure
+// coded object layers.
+func (s xlSets) StorageInfo() StorageInfo {
+	var info StorageInfo
+	info.Backend.Type = Erasure
+	for _, layer := range s.layers {
+		linfo := layer.StorageInfo()
+		info.Total = info.Total + linfo.Total
+		info.Free = info.Free + linfo.Free
+		info.Backend.OnlineDisks = info.Backend.OnlineDisks + linfo.Backend.OnlineDisks
+		info.Backend.OfflineDisks = info.Backend.OfflineDisks + linfo.Backend.OfflineDisks
+	}
+	return info
+}
+
```

Comment:
minor nit - function comments missing for majority of functions in xl-sets.go

Reply:
Fixed